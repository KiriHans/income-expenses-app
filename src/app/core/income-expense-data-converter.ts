import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { IncomeExpense } from './models/income-expenses.model';

export const incomeExpenseConverter: FirestoreDataConverter<IncomeExpense> = {
  toFirestore: (incomeExpense: IncomeExpense) => {
    const { description, amount, type, uid } = incomeExpense;
    return { description, amount, type, uid };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<IncomeExpense>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new IncomeExpense(data.description, data.amount, data.type, data.uid);
  },
};
