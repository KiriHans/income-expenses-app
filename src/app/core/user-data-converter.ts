import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { UserApp } from './models/users.model';

export const userConverter: FirestoreDataConverter<UserApp> = {
  toFirestore: (user: UserApp) => {
    const { displayName, email } = user;
    return { displayName, email };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<UserApp>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new UserApp(data.id, data.displayName, data.email);
  },
};
