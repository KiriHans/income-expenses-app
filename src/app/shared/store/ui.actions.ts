import { createActionGroup, emptyProps } from '@ngrx/store';

export const UiActions = createActionGroup({
  source: 'UI',
  events: {
    'Load UI': emptyProps(),
    'Stop Load UI': emptyProps(),
  },
});
