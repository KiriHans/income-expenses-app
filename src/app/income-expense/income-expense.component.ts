import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';

@Component({
  selector: 'app-income-expense',
  standalone: true,
  imports: [CommonModule],
  templateUrl: './income-expense.component.html',
  styles: [],
})
export class IncomeExpenseComponent {}
