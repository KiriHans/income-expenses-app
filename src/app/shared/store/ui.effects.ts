import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { map, concatMap } from 'rxjs/operators';
import { EMPTY } from 'rxjs';
import { UiActions } from './ui.actions';

@Injectable()
export class UiEffects {
  loadUis$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(UiActions.loadUI),
      concatMap(() =>
        /** An EMPTY observable only emits completion. Replace with your own observable API request */
        EMPTY.pipe(map((data) => UiActions.loadUI()))
      )
    );
  });

  constructor(private actions$: Actions) {}
}
