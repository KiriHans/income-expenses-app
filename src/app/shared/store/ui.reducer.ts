import { createFeature, createReducer, on } from '@ngrx/store';
import { UiActions } from './ui.actions';

export const uiFeatureKey = 'ui';

export interface UIState {
  isLoading: boolean;
}

export const initialState: UIState = {
  isLoading: false,
};

export const uiReducer = createReducer(
  initialState,
  on(UiActions.loadUI, (state): UIState => ({ ...state, isLoading: true })),
  on(UiActions.stopLoadUI, (state): UIState => ({ ...state, isLoading: false }))
);

export const uiFeature = createFeature({
  name: uiFeatureKey,
  reducer: uiReducer,
});
