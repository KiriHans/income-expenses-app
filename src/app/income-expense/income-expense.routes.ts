import { Routes } from '@angular/router';
import { IncomeExpenseComponent } from './components/income-expense.component';
import { DetailsComponent } from './components/details/details.component';
import { StatisticsComponent } from './components/statistics/statistics.component';

export default [
  {
    path: 'income-expenses',
    component: IncomeExpenseComponent,
  },
  {
    path: 'detail',
    component: DetailsComponent,
  },
  {
    path: '',
    pathMatch: 'full',
    component: StatisticsComponent,
  },
] satisfies Routes;
