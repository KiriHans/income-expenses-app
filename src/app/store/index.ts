import { isDevMode } from '@angular/core';
import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import * as ui from '../shared/store/ui.reducer';
import * as auth from './reducers/auth.reducer';
import * as incomeExpense from './reducers/income-expense.reducer';

export const stateFeatureKey = 'state';

export interface AppState {
  [ui.uiFeatureKey]: ui.UIState;
  [auth.authFeatureKey]: auth.State;
  [incomeExpense.incomeExpenseFeatureKey]: incomeExpense.State;
}

export const reducers: ActionReducerMap<AppState> = {
  [ui.uiFeatureKey]: ui.uiFeature.reducer,
  [auth.authFeatureKey]: auth.userReducer,
  [incomeExpense.incomeExpenseFeatureKey]: incomeExpense.reducer,
};

export const metaReducers: MetaReducer<AppState>[] = isDevMode() ? [] : [];
