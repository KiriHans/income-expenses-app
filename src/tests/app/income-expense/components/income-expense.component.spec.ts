import { ComponentFixture, TestBed } from '@angular/core/testing';

import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { Firestore } from '@angular/fire/firestore';
import { IncomeExpenseComponent } from 'src/app/income-expense/components/income-expense.component';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';
import { IncomeExpense } from 'src/app/core/models/income-expenses.model';
import { IncomeExpenseActions } from 'src/app/store/actions/income-expense.actions';

describe('IncomeExpenseComponent', () => {
  let component: IncomeExpenseComponent;
  let fixture: ComponentFixture<IncomeExpenseComponent>;
  let el: DebugElement;
  let compiled: HTMLElement;

  let store: MockStore;

  const initialState = {
    incomeExpense: {
      items: [],
      loading: false,
    },
  } as const;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [IncomeExpenseComponent],
      providers: [provideMockStore({ initialState }), { provide: Firestore, useValue: {} }],
    });
    fixture = TestBed.createComponent(IncomeExpenseComponent);
    component = fixture.componentInstance;
    el = fixture.debugElement;
    compiled = fixture.nativeElement;

    store = TestBed.inject(MockStore);

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should render a form with email and password inputs', () => {
    const title = el.query(By.css('h4'));
    const paragraph = el.query(By.css('p.card-description'));
    const form = el.query(By.css('form'));
    const [description, amount] = el.queryAll(By.css('.form-group input'));
    const type = el.query(By.css('.form-group button'));

    expect(title.nativeElement.textContent).toBe('Add Income / Expenses');
    expect(form.nativeElement.textContent).toBeTruthy();
    expect(paragraph.nativeElement.textContent).toBe('Specify amount and type');

    expect(description.name).toBe('input');
    expect(description.attributes['type']).toBe('text');
    expect(description.attributes['formControlName']).toBe('description');
    expect(description.attributes['placeholder']).toBe('Description');

    expect(amount.name).toBe('input');
    expect(amount.attributes['type']).toBe('number');
    expect(amount.attributes['formControlName']).toBe('amount');
    expect(amount.attributes['placeholder']).toBe('Amount');

    expect(type.name).toBe('button');
    expect(type.nativeElement.textContent.trim()).toBe('Income');
  });

  it("should contain inside each field it's respective label", () => {
    const descriptionLabel = el.query(By.css('.form-group label[for="description"]'));
    const amountLabel = el.query(By.css('.form-group label[for="amount"]'));
    const typeLabel = el.query(By.css('.form-group label[for="type"]'));

    expect(descriptionLabel).toBeTruthy();
    expect(descriptionLabel.name).toBe('label');

    expect(amountLabel).toBeTruthy();
    expect(amountLabel.name).toBe('label');

    expect(typeLabel).toBeTruthy();
    expect(typeLabel.name).toBe('label');
  });

  it('should have a button that is disabled when form is invalid', () => {
    const submitButton = el.query(By.css('button.btn-success'));

    expect(submitButton.nativeElement.disabled).toBeTruthy();
    expect(submitButton.name).toBe('button');
    expect(submitButton.nativeElement.textContent.trim()).toBe('Add');
  });

  it('should have a cancel button that resets the information in the form', () => {
    const cancelButton = el.query(By.css('button[type="reset"]'));

    expect(cancelButton.name).toBe('button');
    expect(cancelButton.nativeElement.textContent.trim()).toBe('Cancel');
  });

  it('should enable button when form is valid', () => {
    const submitButton = el.query(By.css('button.btn-success'));

    component.incomeForm.setValue({
      description: 'Test description',
      amount: 100,
    });

    fixture.detectChanges();

    expect(submitButton.nativeElement.disabled).toBeFalsy();
  });

  it('should change type button when clicked', () => {
    let typeButton = el.query(By.css('.form-group button'));

    expect(typeButton.classes['btn-primary']).toBeTruthy();
    expect(typeButton.nativeElement.textContent.trim()).toBe('Income');

    typeButton.nativeElement.click();

    fixture.detectChanges();

    typeButton = el.query(By.css('.form-group button'));

    expect(typeButton.classes['btn-warning']).toBeTruthy();
    expect(typeButton.nativeElement.textContent.trim()).toBe('Expense');
  });

  it('should change type value when type button is clicked', () => {
    const typeButton = el.query(By.css('.form-group button'));

    expect(component.type).toBe('income');

    typeButton.nativeElement.click();

    expect(component.type).toBe('expense');
  });

  it('should trigger save method when button is pressed', () => {
    jest.spyOn(component, 'save');
    const button = el.query(By.css('button.btn-success'));

    component.incomeForm.setValue({
      description: 'Test description',
      amount: 100,
    });

    fixture.detectChanges();

    button.nativeElement.click();

    expect(component.save).toHaveBeenCalled();
  });

  it('save method should dispatch its action', () => {
    component.incomeForm.setValue({
      description: 'Test description',
      amount: 100,
    });

    fixture.detectChanges();

    const { description, amount } = component.incomeForm.value;
    const incomeExpense = new IncomeExpense(description, amount, component.type);

    jest.spyOn(store, 'dispatch');
    component.save();

    expect(store.dispatch).toHaveBeenCalledWith(
      IncomeExpenseActions.createIncomeExpenses({ incomeExpense })
    );
  });

  it('should show new button with "waiting..." text while the items is being added', () => {
    let button = el.query(By.css('button.btn-success'));

    component.incomeForm.setValue({
      description: 'Test description',
      amount: 100,
    });

    fixture.detectChanges();

    store.setState({
      incomeExpense: {
        items: [],
        loading: true,
      },
    });

    store.refreshState();

    fixture.detectChanges();

    button = el.query(By.css('button.btn-success'));

    expect(button.nativeElement.textContent.trim()).toBe('Waiting...');
  });
});
