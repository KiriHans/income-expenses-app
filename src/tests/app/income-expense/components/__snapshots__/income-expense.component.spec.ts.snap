// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomeExpenseComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="row"
  >
    <div
      class="col-12"
    >
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <h4
            class="card-title"
          >
            Add Income / Expenses
          </h4>
          <p
            class="card-description"
          >
            Specify amount and type
          </p>
          <form
            class="ng-untouched ng-pristine ng-invalid"
            novalidate=""
          >
            <div
              class="form-group"
            >
              <label
                for="description"
              >
                Description
              </label>
              <input
                class="form-control ng-untouched ng-pristine ng-invalid"
                formcontrolname="description"
                name="description"
                placeholder="Description"
                type="text"
              />
              <p
                class="pt-2"
                style="color: rgb(71, 71, 71); font-size: 0.7rem;"
              >
                 It must be less than 100 characters. 
              </p>
            </div>
            <div
              class="form-group"
            >
              <label
                for="amount"
              >
                Amount
              </label>
              <input
                class="form-control ng-untouched ng-pristine ng-invalid"
                formcontrolname="amount"
                name="amount"
                placeholder="Amount"
                type="number"
              />
              <p
                class="pt-2"
                style="color: rgb(71, 71, 71); font-size: 0.7rem;"
              >
                 It must be a positive amount. 
              </p>
            </div>
            <div
              class="form-group"
            >
              <label
                for="type"
              >
                Type
              </label>
              <br />
              <button
                class="btn btn-primary btn-block"
                type="button"
              >
                 Income 
              </button>
              
              
            </div>
            <hr />
            <button
              class="btn btn-success mr-2"
              disabled=""
              type="submit"
            >
              <i
                class="fa fa-save"
              />
               Add 
            </button>
            
            
            <button
              class="btn btn-light"
              type="reset"
            >
              Cancel
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
